### DISCUSSION TITLE
HTML Modules

### DISCUSSION THREAD

**Discussion by MzHub** [Devographics/surveys#169] [State of HTML Features]
Import HTML files via JS imports and access their elements and JS exports.

### Code sample

```html
<script type="module">
	import { TabList } from "./tablist.html";
	customElements.define("tab-list", TabList);
</script>
```

[`Main.html`](https://github.com/WICG/webcomponents/blob/c351fc9b4a40e04cb7a9c7ba4f3213a13c0f559e/proposals/HTML-Modules-Demo/Main.html)
```html
<!doctype html>
<html>
    <head>
        <title>HTML Modules Demo</title>
        <script type="module">
            import { HTML5Element } from "./html5Element.html";
            window.customElements.define("my-html5-element", HTML5Element);
        </script>
    </head>
    <body>
        <my-html5-element></my-html5-element>
    </body>
</html>
```

[`HTML5-Element.html`](https://github.com/WICG/webcomponents/blob/c351fc9b4a40e04cb7a9c7ba4f3213a13c0f559e/proposals/HTML-Modules-Demo/HTML5-Element.html)

<details>

<summary>Click to show</summary>

```html
<template id="html5ElementTemplate">
  <style>
      .outerDiv {
          border:0.1em solid blue;
          display:inline-block;
          padding: 0.4em;
      }

      .devText {
          font-weight: bold;
          font-size: 1.4em;
          text-align: center;
          margin-top: 0.3em;
      }

      .mainImage {
          height:254px;
      }
  </style>
  <div class="outerDiv">
      <img class="mainImage" src="https://www.w3.org/html/logo/downloads/HTML5_Logo_512.png" />
      <div class="devText">HTML Modules Are Great!</div>
  </div>
</template>
<script type="module">
  let moduleDocument = import.meta.document;

  class HTML5Element extends HTMLElement {
      constructor() {
          super();
          console.log("HTML5Element constructor");
          let shadowRoot = this.attachShadow({ mode: "closed" });
          let template = moduleDocument.getElementById("html5ElementTemplate");
          shadowRoot.appendChild(document.importNode(template.content, true));
      }

      connectedCallback() {
          console.log("HTML5Element connectedCallback");
      }
  }

  export { HTML5Element };
</script>
```

</details>

### Specification URL:

https://github.com/WICG/webcomponents/blob/gh-pages/proposals/html-modules-explainer.md

### Any other links?

https://chromestatus.com/feature/4854408103854080
https://github.com/MicrosoftEdge/MSEdgeExplainers/blob/main/HTMLModules/designDoc.md

### How stable is this feature?

Future (No implementations yet)

### Type of feature

JS API

### What focus area does this feature fall under?

Consuming Web Components, Making Web Components, Templating

### Details

HTML Modules is a big deal for me personally.

I would love to use the native `<template>` element for my Web Components/Custom Elements, without placing all the `<template>` tags for different components in the website's main document itself. HTML Modules is the only native way I've found to do this.

If I understand it correctly, it would allow components to be `.html` files with a `<template>`, `<script>` and maybe `<style>` as well, and you could then easily import the whole component.

**Comment by MzHub** [comment]
HTML Modules is a big deal for me personally.

I would love to use the native `<template>` element for my Web Components/Custom Elements, without placing all the `<template>` tags for different components in the website's main document itself. HTML Modules is the only native way I've found to do this.

If I understand it correctly, it would allow components to be `.html` files with a `<template>`, `<script>` and maybe `<style>` as well, and you could then easily import the whole component.

**Reply by MzHub** [reply]
Here's a code example from a project with a partial polyfill for HTML Modules:

- [Import example](https://github.com/indigane/hajalista/blob/ca6a09a/index.html#L36-L39)
- [Checkbox list item component example](https://github.com/indigane/hajalista/blob/ca6a09a/checklist-item.html)
- [Partial polyfill](https://github.com/indigane/hajalista/blob/ca6a09a/js/hmfill.js)

**Comment by LeaVerou** [comment]
Yeah, I was wondering about this. There was a lot of buzz around it, but then Safari (IIRC) vetoed it and it was eventually removed even by browsers that supported it.

**Reply by MzHub** [reply]
I believe that was HTML Imports. This is similar, but slightly different, the explainer goes into more detail on how they differ: https://github.com/WICG/webcomponents/blob/c351fc9/proposals/html-modules-explainer.md#why-are-html-modules-needed

**Reply by LeaVerou** [reply]
You are absolutely correct, I was talking about HTML Imports! I've updated your first post with a code sample (to ensure others don't also get confused) and entered your description in "more details", since it was empty.

**Reply by MzHub** [reply]
Thanks!

If you don't mind, I'll change the code sample to this [HTML Modules Demo](https://github.com/WICG/webcomponents/blob/c351fc9b4a40e04cb7a9c7ba4f3213a13c0f559e/proposals/HTML-Modules-Demo/Main.html) from the proposal, which I think is a bit clearer on how it would work.
